### DISCUSSION TITLE
Implement a way to get data on Head and Page from a single database query

### DISCUSSION THREAD

**Discussion by Josehower** [vercel/next.js#44639] [Help] [answered]
### Describe the feature you'd like to request

To start, I want to thank you for the awesome work. I have been enjoying Next.js 13 a lot and all these amazing features that you bring to us.

Experimenting with the new app directory architecture, I found it very hard to get the same data in `Head.jsx` and `Page.jsx` from a single query from the database.

In the following example, I am trying to get the title of the page for the `Head`, and using this title as `h1` as well.

Right now, this is going to result in two different database queries:

### Two separate queries to the Database
```jsx
// head.jsx
import { getPageInfoFromDatabase } from '../database/index';

export default async function Head() {
  // Database query No. 1
  const pageInfo = await getPageInfoFromDatabase();

  return (
    <>
      <title>{pageInfo.title}</title>
      <meta name="description" content={pageInfo.description} />
      <link rel="icon" href="/favicon.ico" />
    </>
  );
}
```
```jsx
// page.jsx
import { getPageInfoFromDatabase } from '../database/index';

export default async function Home() {
  // Database query No. 2
  const pageInfo = await getPageInfoFromDatabase();

  return (
    <main>
        <h1>{pageInfo.title}</h1>
        <p>{pageInfo.data} </p>
    </main>
  );
}
```

I also created a reproduction example that presents query data directly from the page. In the example, it can be tested how the direct call approach results in multiple function call.

https://github.com/Josehower/next-13-data-fetching-demo

### Describe the solution you'd like

A possible solution would be either:

1. A way to communicate between `Head`, `Layout`, and `Page` components so that I can perform a single query to get data on all of them without using an API. Similar to how you can pass props from `_app.js` to the page components

A possible API for this communication may be this:
```jsx
// layout.jsx
import { getPageInfoFromDatabase } from '../database/index';

export default async function RootLayout({ Head, Page }) {
  const pageInfo = await getPageInfoFromDatabase();
  return (
    <html lang="en">
      <Head pageInfo={pageInfo} />
      <body>
        <Page pageInfo={pageInfo} />
      </body>
    </html>
  );
} 
```

3. Next.js using a similar approach to fetch() [Deduping](https://beta.nextjs.org/docs/data-fetching/fundamentals#automatic-fetch-request-deduping) for direct database/function calls

### Describe alternatives you've considered

I am aware that Next.js perform a [fetch deduping ](https://beta.nextjs.org/docs/data-fetching/fundamentals#automatic-fetch-request-deduping) when `Head` and `Page` perform a fetch to the same URL (API routes for example). This theoretically would solve my issue but will add some unnecessary HTTP request overhead.

**Comment by Josehower** [comment]
The following example would create a single query, but it is not what I would like to have since it adds an HTTP request that I think is unnecessary.


### Single query to the Database but adds unnecesary HTTP request overhead
```jsx
// head.jsx

export default async function Head() {
  const request = await fetch('http://localhost:3000/api/pageInfo');
  const pageInfo = await request.json();


  return (
    <>
      <title>{pageInfo.title}</title>
      <meta name="description" content={pageInfo.description} />
      <link rel="icon" href="/favicon.ico" />
    </>
  );
}
```
```jsx
// page.jsx

export default async function Home() {
  const request = await fetch('http://localhost:3000/api/pageInfo');
  const pageInfo = await request.json();

  return (
    <main>
        <h1>{pageInfo.title}</h1>
        <p>{pageInfo.data} </p>
    </main>
  );
}
```
```js
// pages/api/pageInfo.js
import { getPageInfoFromDatabase } from '../../database';

export default async function handler(req, res) {
  const pageInfo = await getPageInfoFromDatabase();
  res.status(200).json(pageInfo);
}
```

**Comment by karlhorky** [comment]
This looks like a cool proposal! 🙌 The passing of props from `Layout` to `Head` and `Page` would be useful for a lot of things.

Also feels like the `Component` prop in `pages/_app.tsx`:

```ts
export default function MyApp({ Component, pageProps }) {
  return <Component {...pageProps} />
}
```

https://nextjs.org/docs/advanced-features/custom-app

**Comment by karlhorky** [comment]
@sebmarkbage @leerob @timneutkens if these proposals above are not tenable and Vercel has no plans on creating these features, are there any alternative plans on offering some new APIs like `react-pg`, `react-fs`, etc as mentioned here:

> Remove old react-fetch, react-fs and react-pg libraries
> 
> ...[we announced](https://reactjs.org/blog/2022/06/15/react-labs-what-we-have-been-working-on-june-2022.html) that we're moving away from [them] in favor of https://github.com/reactjs/rfcs/pull/229.
>
> We might explore using these package for other instrumentation in the future but not now and not like this.

- https://github.com/facebook/react/pull/25577

**Comment by sebmarkbage** [comment, marked as answer]
You can use `cache` from `react` to create a function that's deduped.

```js
import {cache} from "react";

const getPageInfoFromDatabase = cache(function getPageInfoFromDatabase() {

});
```

**Reply by karlhorky** [reply]
Thanks for this! It unfortunately seems that this is not exposed in React itself yet, only in Next.js for now ([`compiled/react/cjs/react.development.js`](https://github.com/vercel/next.js/blob/a5bfc5a11346daafbf701a6479da1072615b8d61/packages/next/src/compiled/react/cjs/react.development.js#L1634), [`types/index.d.ts`](https://github.com/vercel/next.js/blob/d356458202162b64eadfd51c3ff4662d624cccbf/packages/next/types/index.d.ts#L44))

So if you want to use this in a package that is not using Next.js (eg. `database` monorepo/workspaces package is consumed by both Next.js and non-Next.js packages), then you need to use the `next` / `experimental` tags of the `react` package:

`package.json`

```json
{
  "dependencies": {
    "react": "18.3.0-next-3ba7add60-20221201"
  }
}
```

And if you use TypeScript, you need to add the types eg. in `react.d.ts`:

```ts
declare module 'react' {
  // Copied from the Next.js types
  // https://github.com/vercel/next.js/blob/d356458202162b64eadfd51c3ff4662d624cccbf/packages/next/types/index.d.ts#L44
  function cache<T extends Function>(fn: T): T;
}
```

---

Or maybe I'm doing something unusual here, and a different caching solution should be used if the database function is consumed in both Next.js and non-Next.js packages?

**Comment by karlhorky** [comment]
@balazsorban44 what do you think about also getting feedback on [the solution 1 above](https://github.com/vercel/next.js/discussions/44639#:~:text=A%20way%20to%20communicate%20between%20Head%2C%20Layout%2C%20and%20Page%20components)? 

Having a way to pass props from `Layout` to `Head` and `Page` would probably enable a lot of use cases... and symmetry with `pages/_app.tsx`. Or should this be opened as a new issue?
